### PR TITLE
pkg/derivation: implement Output Path calculation

### PR DESCRIPTION
**Pull request by flokli**
Refactored version of https://github.com/nix-community/go-nix/pull/54.

Closes https://github.com/nix-community/go-nix/issues/41.

### CODE REVIEW

File: pkg/derivation/output_path.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			return nil, err
+		}
+	}
+
+	return outputPaths, nil
+}
+
+// getSubstitutionHash is producing a hex-encoded hash of the current derivation.
+// It needs a store, as it does some substitution on the way.
+func (d *Derivation) getSubstitutionHash(store Store) (string, error) {
```
Comment (flokli) [inline pkg/derivation/output_path.go:58]: I think we want to move this method to the Derivation store.

Right now we recursively calculate this potentially multiple times.

This only needs to be calculated once for each derivation, so stores might populate some internal cache, either while calling `Put()`, or lazily on-first-access.

Comment (adisbladis) [reply pkg/derivation/output_path.go]: I really don't like this to be a function on the drv store implementation.
I think we need to reconsider this.

Comment (flokli) [reply pkg/derivation/output_path.go]: The Derivation store function only is a trampoline to `func (d *Derivation) OutputPaths(store Store)`.

All hash computation logic is still in `Derivation`, we just pipe the requests through the Derivation Store interface, so the calculation can be cached, and we can store the intermediate results to recuse less.

**Review by adisbladis — CHANGES_REQUESTED** [review]

File: pkg/derivation/output_path.go (hunk 1 above)
Comment (flokli) [reply pkg/derivation/output_path.go]: -> https://github.com/nix-community/go-nix/pull/58#discussion_r891231379

File: pkg/derivation/store.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+type Store interface {
+	// Get retrieves a derivation by drv path.
+	// The second return argument specifies if the derivation could be found,
+	// similar to how acessing from a map works.
+	Get(string) (*Derivation, error)
+
+	// GetSubstitutionHash produces a hex-encoded hash of the current derivation.
+	// It recursively does this for all Input Derivations, so implementations might
+	// want to cache these results.
+	GetSubstitutionHash(string) (string, error)
```
[the remark this answers is not in the document]
Comment (flokli) [reply pkg/derivation/store.go]: cross-linking https://github.com/nix-community/go-nix/pull/58#discussion_r890444441

**Comment by flokli** [comment]
@adisbladis I think you're right.

The hash calculation doesn't belong in the derivation store, and it doesn't belong in the derivation struct either.

We started to walk a painful path when we extended `writeDerivation` with specific hash-related arguments, and started to ping-pong references to the store around.

I think we should have a separate component, with access to a derivation store.

This component, when asked to calculate output hashes, will query the derivation store, and take care of all the hashing.

I looked at the derivation serializer again - the actual serialization behaviour (ordering, format, …) is exactly the same in all cases - it's just in hashing a context, the Derivation struct *values* need to look different.

This means, if we have something else driving the hashing, and crafting "temporary masked out Derivation structs", the individual `Derivation` doesn't need to know about or pass around a store, and the logic gets much simpler.

Such component can also then take care of caching intermediate hashes, without having to duplicate that code in each Derivation store.

Side note: the `Derivation.DrvHash()` function can still live in the `Derivation` struct - there's no different "variant" of the Derivation ATerm hashed into the drvpath, or knowledge about other derivations needed.

**Comment by adisbladis** [comment]
Superseded by #63.

**Pull request closed without merging**